### PR TITLE
[SPARK-7190] [SPARK-8804] [SPARK-7815] [SQL] unsafe UTF8String

### PR DESCRIPTION
**Pull request by davies**
Let UTF8String work with binary buffer. Before we have better idea on manage the lifecycle of UTF8String in Row, we still do the copy when calling `UnsafeRow.get()` for StringType.

cc @rxin @JoshRosen

### CODE REVIEW

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 1]
```diff
@@ -90,13 +87,15 @@ public int numBytes(final byte b) {
    */
   public int length() {
```
Comment (rxin) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: not related to this pr, but it'd be less confusing if we rename this numChars

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 2]
```diff
@@ -35,7 +34,9 @@
 public final class UTF8String implements Comparable<UTF8String>, Serializable {
 
   @Nonnull
-  private byte[] bytes;
+  private final Object base;
+  private final long offset;
+  private final int size;
```
Comment (rxin) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: size -> numBytes to make it more clear

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 3]
```diff
@@ -209,6 +242,16 @@ public boolean equals(final Object other) {
 
   @Override
   public int hashCode() {
-    return Arrays.hashCode(bytes);
+    long result = 1;
```
Comment (tarekauel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Maybe it's a stupid question, but why do we start with `result = 1` instead of `0`, compared to http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/lang/String.java#1493 and http://grepcode.com/file/repository.cloudera.com/content/repositories/releases/org.apache.hadoop.hive/hive-exec/0.7.1-cdh3u1/org/apache/hadoop/hive/ql/udf/generic/GenericUDFHash.java#47

Do we want to cache the hash value like OpenJdk?

Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Just copied from Arrays.hashCode(), I think it doesn't matter much.

For caching, we didn't really use UTF8String as the key, it's often to use InternalRow as key, so it's better to cache the hashCode inside InternalRow.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 1 above)
Comment (rxin) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: actually while you are at this, let's rename it to prevent misusing this.

Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: This is to match with String.length(), should not it?

Comment (rxin) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Yea I think it's best to not have it named length because it is a little bit ambiguous here. It could mean length of the string in terms of characters, or length in terms of number of bytes.

Comment (rxin) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: In Java string it is not ambiguous because people in general don't think about string in terms of a byte array

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 4]
```diff
@@ -106,92 +105,126 @@ public int length() {
...
-    if (prefix.length + offsetInBytes > bytes.length || offsetInBytes < 0) {
+  private long getLong(int i) {
+    return UNSAFE.getLong(base, offset + i);
+  }
+
+  private byte getByte(int i) {
+    return UNSAFE.getByte(base, offset + i);
+  }
+
+  private boolean matchAt(final UTF8String s, int pos) {
```
Comment (JoshRosen) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: The loops here in `matchAt`, could probably be replaced by a single call to `ByteArrayMethods.arrayEquals`, provided that we make a non-word-aligned version of that method.  There is some example code in Guava that you could base this on. We can defer this for later, too, but I figure that since we've basically already written this method here then we might as well put it into a place where it can be used more generically.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 5]
```diff
@@ -263,17 +263,17 @@ public Object get(int i) {
...
-        PlatformDependent.BYTE_ARRAY_OFFSET,
-        size
-      );
       if (isString) {
-        return UTF8String.fromBytes(bytes);
+        return new UTF8String(baseObject, baseOffset + offset, size);
       } else {
+        final byte[] bytes = new byte[size];
+        PlatformDependent.copyMemory(
+                baseObject,
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: This looks over-indented.

Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: Intellij did, will fix it

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 6]
```diff
@@ -106,92 +105,126 @@ public int length() {
...
+    // this is locale aware
     return UTF8String.fromString(toString().toLowerCase());
   }
 
   @Override
   public String toString() {
     try {
-      return new String(bytes, "utf-8");
+      // this is slow
+      return new String(getBytes(), "utf-8");
```
Comment (JoshRosen) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Just curious: why do we need the copy here in `getBytes()`?  Does String hold onto the byte array or otherwise manipulate it?

Comment (JoshRosen) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: i.e.. why not pass `bytes` directly?

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 7]
```diff
@@ -45,33 +46,29 @@
     6, 6, 6, 6};
 
   public static UTF8String fromBytes(byte[] bytes) {
```
Comment (JoshRosen) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Maybe it's not necessary, but should we add a comment to call out the fact that this doesn't defensively copy `bytes`?

File: unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java [hunk 8]
```diff
@@ -25,21 +25,30 @@
...
+    Assert.assertEquals(s1, s2);
 
-    Assert.assertEquals(UTF8String.fromString(str).hashCode(),
-      UTF8String.fromBytes(str.getBytes("utf8")).hashCode());
+    Assert.assertEquals(s1.hashCode(), s2.hashCode());
+
+    Assert.assertEquals(s1.compare(s2), 0);
+    Assert.assertEquals(s1.compareTo(s2), 0);
+
+    Assert.assertEquals(s1.contains(s2), true);
```
Comment (JoshRosen) [inline unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java]: I think you can use `assertTrue` here.  As long as you're changing this file, you might as well change it to statically import the `assert*` methods.

File: unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java [hunk 9]
```diff
@@ -25,21 +25,30 @@
...
+    Assert.assertEquals(s2.contains(s1), true);
+    Assert.assertEquals(s1.startsWith(s1), true);
+    Assert.assertEquals(s1.endsWith(s1), true);
   }
 
   @Test
   public void basicTest() throws UnsupportedEncodingException {
     checkBasic("hello", 5);
-    checkBasic("世 界", 3);
+    checkBasic("大 千 世 界", 7);
```
Comment (JoshRosen) [inline unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java]: We should probably also add a test case for the empty string.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 10]
```diff
@@ -201,7 +234,7 @@ public int compare(final UTF8String other) {
   @Override
   public boolean equals(final Object other) {
     if (other instanceof UTF8String) {
-      return Arrays.equals(bytes, ((UTF8String) other).getBytes());
+      return compareTo((UTF8String) other) == 0;
```
Comment (JoshRosen) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Since I suspect that string equality comparisons could be a very frequent / expensive operation, this might be a case where it would be worthwhile to use a fast byte array equality method (see my suggestion upthread on taking the byte comparison loop in `matches` and factoring it out into a static method in `ByteArrayMethods`).  It might also be faster to express this as a check that the strings have the same length and that one string matches at the start of the other.

**Comment by JoshRosen** [comment]
This looks pretty good to me overall; I left a few small optimization related comments.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 11]
```diff
@@ -44,60 +47,77 @@
...
+    this.offset = offset;
+    this.numBytes = size;
   }
 
   /**
    * Returns the number of bytes for a code point with the first byte as `b`
    * @param b The first byte of a code point
    */
-  public int numBytes(final byte b) {
+  public int numBytesForFirstByte(final byte b) {
```
Comment (tarekauel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: This could method could be static

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 6 above)
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: We don't have `bytes` (byte[]) right now.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 4 above)
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: done

**Comment by davies** [comment]
@JoshRosen @rxin Please take another look.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 12]
```diff
@@ -209,6 +308,10 @@ public boolean equals(final Object other) {
 
   @Override
   public int hashCode() {
-    return Arrays.hashCode(bytes);
```
Comment (JoshRosen) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: If performance is important here, we could consider adding a `hashUnsafeBytes` method to `Murmur3_x86_32` and could use that instead.  We can also defer that to a followup PR, since I'd like to get this merged soon.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 13]
```diff
@@ -106,92 +168,125 @@ public int length() {
...
+      }
+      i += n;
+    }
+
+    return fromBytes(buf);
   }
 
   public UTF8String toLowerCase() {
-    return UTF8String.fromString(toString().toLowerCase());
+    // It's always have the same number of bytes for upper case
```
Comment (JoshRosen) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: "for lower case"

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 14]
```diff
@@ -106,92 +168,125 @@ public int length() {
...
   public UTF8String toUpperCase() {
-    return UTF8String.fromString(toString().toUpperCase());
+    // It's always have the same number of bytes for upper case
+    byte[] buf = new byte[numBytes];
+    copyMemory(base, offset, buf, BYTE_ARRAY_OFFSET, numBytes);
+
+    for (int i = 0; i < numBytes; ){
+      int n = numBytesForFirstByte(getByte(i));
+      int code = codePointAt(i, n);
+      int upper = Character.toUpperCase(code);
```
Comment (JoshRosen) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: According to the Javadoc for [Character.toUpperCase()](https://docs.oracle.com/javase/7/docs/api/java/lang/Character.html#toUpperCase(int)):

> In general, String.toUpperCase() should be used to map characters to uppercase. String case mapping methods have several benefits over Character case mapping methods. String case mapping methods can perform locale-sensitive mappings, context-sensitive mappings, and 1:M character mappings, whereas the Character case mapping methods cannot.

Based on this, it sounds like our implementations of `toUpperCase()` and `toLowerCase()` could potentially behave differently than their `String` equivalents.  Do you have any examples of inputs where this is the case?  Is this something that we should be concerned about?

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 6 above)
Comment (JoshRosen) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Ah, right, since this is a pointer.  My bad.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 15]
```diff
@@ -106,92 +168,125 @@ public int length() {
...
 
   @Override
   public int compareTo(final UTF8String other) {
-    final byte[] b = other.getBytes();
-    for (int i = 0; i < bytes.length && i < b.length; i++) {
-      int res = bytes[i] - b[i];
+    int len = numBytes < other.numBytes ? numBytes : other.numBytes;
+    int i = 0;
+    // TODO: compare 8 bytes as unsigned long
+    while (i < len) {
```
Comment (JoshRosen) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Is there an advantage to using a `while` loop here instead of a `for` loop?

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 16]
```diff
@@ -106,92 +168,125 @@ public int length() {
...
+      int res = (getByte(i) & 0xFF) - (other.getByte(i) & 0xFF);
```
Comment (JoshRosen) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Maybe we should add a comment here to explain that this is performing unsigned comparison.

File: unsafe/src/main/java/org/apache/spark/unsafe/array/ByteArrayMethods.java [hunk 17]
```diff
@@ -38,18 +38,24 @@ public static int roundNumberOfBytesToNearestWord(int numBytes) {
    * Optimized byte array equality check for 8-byte-word-aligned byte arrays.
```
Comment (JoshRosen) [inline unsafe/src/main/java/org/apache/spark/unsafe/array/ByteArrayMethods.java]: This comment is now out-of-date and should be updated to reflect the fact that the new method does not require the input arrays' lengths to be multiples of 8 bytes.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 14 above)
Comment (JoshRosen) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Using ScalaCheck, I think I found some failing test cases.  I'll open a PR against this to demonstrate in a sec.

**Comment by JoshRosen** [comment]
I think that ScalaCheck found a bug in `substring(0, s.length)`:

```scala
scala> val s = "ߵ梷"
s: String = ߵ梷

scala> s.substring(0, s.length) == s
res0: Boolean = true

scala> import org.apache.spark.unsafe.types.UTF8String
import org.apache.spark.unsafe.types.UTF8String

scala> UTF8String.fromString(s).substring(0, s.length).toString == s
res1: Boolean = false

scala> UTF8String.fromString(s).substring(0, s.length).toString
res2: String = ߵ�

scala>
```

**Comment by JoshRosen** [comment]
Some experiments with property-based checks here: https://github.com/JoshRosen/spark/commit/2595ec25ed4d8d3e4172eb5e2e9f13b4f876cfb4

Here's an problem that it found with `toLowerCase`:

```scala
scala> val s = "Ề"
s: String = Ề

scala> s.toLowerCase
res0: String = ề

scala> UTF8String.fromString(s).toLowerCase
res1: org.apache.spark.unsafe.types.UTF8String = Ề
```

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 12 above)
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Once we change the hashCode of String, some PySpark doc tests will fail, so I revert the change, let's do it in separate PR.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 14 above)
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Good catch, will fix it.

**Comment by JoshRosen** [comment]
@davies, I tested out your latest patch for `toUpperCase` / `toLowerCase` and it doesn't seem to have fixed the example that I listed above.

**Comment by JoshRosen** [comment]
FYI, this is still failing the toUpper/Lower bug mentioned in https://github.com/apache/spark/pull/7197#issuecomment-119015425; just wanted to leave a comment so we don't merge it prematurely w/o deciding whether to fix this or defer to a followup / choose to adopt different semantics.

**Comment by davies** [comment]
@JoshRosen Thanks for the comments, two bugs found!

**Comment by davies** [comment]
@JoshRosen One comment about the String.length, it's not the number of characters ( but code point units instead), so we can't use `UTF8String.fromString(s).substring(0, s.length)`. UTF8String.numChars() is the number of characters. So `strlen` will have different results than Hive (for some special characters). cc @rxin

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 18]
```diff
@@ -106,92 +165,152 @@ public int length() {
...
-    return startsWith(prefix.getBytes(), 0);
+    return matchAt(prefix, 0);
   }
 
   public boolean endsWith(final UTF8String suffix) {
-    return startsWith(suffix.getBytes(), bytes.length - suffix.getBytes().length);
+    return matchAt(suffix, numBytes - suffix.numBytes);
   }
 
+  private static String lang = Locale.getDefault().getLanguage();
```
Comment (chenghao-intel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Add 'final`, and the same for `localeDependent`.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 19]
```diff
@@ -106,92 +165,152 @@ public int length() {
...
+      return fromString(toString().toLowerCase());
+    }
+
+    byte[] buf = null;
+    for (int i = 0; i < numBytes; ){
+      int n = numBytesForFirstByte(getByte(i));
+      int code = codePointAt(i, n);
+      int upper = Character.toUpperCase(code);
+      if (upper != code) {
+        if (localeDependent) {
```
Comment (chenghao-intel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: This was checked in line 230, need to skip it the checking.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 6 above)
Comment (chenghao-intel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: `String` actually has the constructor like:
```
public String(byte bytes[], int offset, int length, String charsetName)
```
Will that be helpful to avoid the possible byte array copying inside of the `getBytes`?

**Comment by chenghao-intel** [comment]
@davies @JoshRosen I have one question for expression in general, since all of the intermediate objects are immutable in expression evaluation, we have to keep creating new object like `UTF8String` instances for every iteration. Probably it causes big overhead in a deep expression tree. Will that be possible to make the mutable object instead? Just like the `org.apache.hadoop.io.Writable`, so we can reuse the intermediate object.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 6 above)
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: When we do the copy in `getBytes`, `base` could be not a `byte[]`, so we still need the copy. There is a case that we could avoid the copy, `base` is `byte[]`, and `offset` is not pointed to the first byte, do we still need to optimize this case? We can do it later.

**Comment by davies** [comment]
@chenghao-intel Before we make UTF8String mutable, we should have clear idea that how to manage the lifecycle of UTF8String (when do the copying). Right now, we didn't do deep copy for Row, that's the reason I haven't enable use pointer instead of copy for `UnsafeRow.get()`.

**Comment by chenghao-intel** [comment]
Ok, I just a little bit concern about the performance, as we always create objects like `UTF8String` for every expression evaluation.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 20]
```diff
@@ -106,102 +132,203 @@ public int length() {
...
+      i += n;
+    }
+    return buf != null ? fromBytes(buf) : this;
   }
 
   @Override
   public String toString() {
     try {
-      return new String(bytes, "utf-8");
+      return new String(getBytes(), "utf-8");
```
Comment (chenghao-intel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: String has the constructor like `public String(byte bytes[], int offset, int length, String charsetName)`, will it be faster if we pass the bytes directly? as In `getBytes`, it's possible to duplicate the byte array in some case.

**Comment by scwf** [comment]
@davies  here is a bug when this PR is in,  that is
when set executor memory >= 32g, all the queries for string field  will have problem. seems it return empty/garbled string to user.

I am not familiar with the unsafe utf8string and not fund the root cause.

**Comment by davies** [comment]
@scwf Can you create a JIRA and provide a way to reproduce your problem?

**Comment by scwf** [comment]
@davies https://issues.apache.org/jira/browse/SPARK-9725
